connaisseur/alert.py

```diff
@@ -0,0 +1,177 @@
+        self.throw_if_alert_sending_fails = receiver_config.get(
+            "fail_if_alert_sending_fails", True
+        )
+        self.payload = self._construct_payload(receiver_config)
+        self.headers = get_headers(receiver_config)
+
+    def _construct_payload(self, receiver_config):
+        try:
+            with open(
+                "{}/{}.json".format(os.getenv("ALERT_CONFIG_DIR"), self.template),
+                "r",
+            ) as templatefile:
+                template = json.load(templatefile)
+        except Exception as err:
+            raise ConfigurationError(
+                "Template file for alerting payload is either missing or invalid JSON: {}".format(
+                    str(err)
+                )
+            )
+        payload = sanitize(template, self.context)
+        if receiver_config.get("payload_fields") is not None:
+            payload.update(receiver_config.get("payload_fields"))
+        return payload
+
+    def send_alert(self):
+        try:
+            response = requests.post(
+                self.receiver_url, data=self.payload, headers=self.headers
+            )
+            response.raise_for_status()
+            logging.info("sent alert to %s", self.template.split("_")[0])
+        except Exception as err:
+            if self.throw_if_alert_sending_fails:
+                raise AlertSendingError(str(err))
+            logging.error(err)
+        return response
+
+
+def load_config():
+    with open(
+        "{}/alertconfig.json".format(os.getenv("ALERT_CONFIG_DIR")), "r"
+    ) as configfile:
+        return json.loads(configfile.read())
+
+
+def get_images(admission_request):
+    relevant_spec = get_container_specs(
+        admission_request.get("request", {}).get("object", {})
+    )
+    return [container.get("image") for container in relevant_spec]
+
+
+def get_headers(receiver_config):
+    headers = {"Content-Type": "application/json"}
+    additional_headers = receiver_config.get("custom_headers")
+    if additional_headers is not None:
+        for header in additional_headers:
+            key, value = header.split(":", 1)
+            headers.update({key.strip(): value.strip()})
+    return headers
+
+
+def send_alerts(admission_request, *, admitted):
+    alert_config = load_config()
+    event_category = "admit_request" if admitted else "reject_request"
+    if alert_config.get(event_category) is not None:
+        for receiver in alert_config[event_category]["templates"]:
+            message = (
+                alert_config[event_category].get(
+                    "message", "CONNAISSEUR admitted a request."
+                )
+                if admitted
+                else alert_config[event_category].get(
+                    "message", "CONNAISSEUR rejected a request."
+                )
+            )
+            alert = Alert(message, receiver, admission_request)
+            alert.send_alert()
+
+
+def sanitize(template_dict, context):
```

Comment:
is `sanitize` the right name here? if i understand this correctly, the function renders all the values in the given dict, replacing all placeholders with the values from `context`. ofcourse internally some sanitization will be done, but that's not the main purpose of this function. i'd rename it to something like `render_template`.